---
title: "ATAC-seq analysis (for GO and clustering)"
author: "Tram (edited from Jin's version)"
date: "2025-04-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## ATAC-seq

```{r, messgae=F, include=F}
set.seed(2025)
options(max.print = 1e7)

library(EnvStats)
library(MASS)
library(knitr)
library(tidyr)
library(ggplot2)
library(expm)
library(Matrix)
library(emmeans)
library(UsingR)
library(dplyr)
library(devtools)
library(rafalib)
library(GSE5859Subset)
library(genefilter)
library(qvalue)
library(GenomicRanges)
library(GenomicAlignments)
library(DESeq2)
library(limma)
library(edgeR)
library(reshape2)
library(pander)
library(Hmisc)
library(pastecs)
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(clusterProfiler)
library(ReactomePA)
library(pheatmap)
library(MAnorm2)
library(AnnotationDbi)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(EnhancedVolcano)
library(Cairo)
library(dendextend)

```

### Data Preprocessing

```{r}
# Load Data
atac_data <- read.table('ATAC-seq-data.xls', header = TRUE)

# Remove all the rows where not a single sample has more than 50 reads.
# refer to: https://tobiasrausch.com/courses/atac/atac-seq-data-analysis.html#removing-missing-peaks
atac_data <- na.omit(atac_data)
atac_data <- atac_data[apply(atac_data[, 4:ncol(atac_data)], 1, max) > 50,]
```

```{r}
# Convert the data into a GRanges object
peaks_gr <- GRanges(seqnames = atac_data$chrom,
                    ranges = IRanges(start = atac_data$start , end = atac_data$end))

# Obtain gene annotations and define promoter regions (TSS Â±3000 bp)
genes_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene)
promoter_gr <- promoters(genes_gr, upstream = 3000, downstream = 3000)

# Find overlaps between ATAC-seq peaks and promoter regions
peaks_in_promoters <- subsetByOverlaps(peaks_gr, promoter_gr)
overlap_hits <- findOverlaps(peaks_gr, promoter_gr)
filtered_peaks <- peaks_in_promoters
```

```{r}
df_promoters <- as.data.frame(filtered_peaks)
head(df_promoters)
names(df_promoters)[names(df_promoters) == "seqnames"] <- "chrom"
head(df_promoters)
df_merged <- merge(atac_data, df_promoters, by = c("chrom", "start", "end"))
head(df_merged)
atac <- df_merged
atac[, 4:19] <- atac[, 4:19] + 1
```


### Data Normalization

```{r}
# refer to: https://hds-sandbox.github.io/bulk_RNAseq_course/develop/05c_count_normalization.html

# Define design matrix
counts <- atac[, 4:11]
design <- data.frame(row.names = colnames(atac[, 4:11]),
                     celltype = factor(rep(c("HSC", "CMP", "CFUE", "Ery"), each = 2)))
# Create DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = design,
                              design = ~celltype)
dds <- DESeq(dds)
# Normalize data
cm <- data.frame(counts(dds, normalized=TRUE))
rownames(cm) <- paste0(atac$chrom, '_', atac$start, '_', atac$end)
```


#### Analysis for each pair

```{r}
# intersected with promoters
design <- data.frame(row.names = colnames(atac[,c(6,7,8,9)]),
                     celltype = factor(rep(c("CMP", "CFUE"), each = 2)))
dds <- DESeqDataSetFromMatrix(countData = atac[,c(6,7,8,9)],
                              colData = design,
                              design = ~celltype)

dds_chr1 <- DESeq(dds,quiet = TRUE)
res <- results(dds_chr1)
print(sum(res$padj < 0.01 & abs(res$log2FoldChange) > 1))

CairoPNG("MA_pair6.png", width = 600, height = 350)
DESeq2::plotMA(res, ylim = c(-5, 5), main = "MA Plot: Differential Chromatin Patterns (CMP vs. CFUE)")
dev.off()

CairoPNG("vol_pair6.png", width = 600, height = 400)
ggplot(data = res, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(alpha = 0.5, size = 2, aes(color = ifelse(padj < 0.05, "Significant", "Not significant"))) +
  scale_color_manual(values = c("black", "red"), guide = FALSE) +
  theme_minimal() +
  labs(x = "Log2 Fold Change (CMP vs. CFUE)", y = "-log10(Adjusted p-value)") +
  ggtitle("Volcano Plot: Differential Chromatin Patterns (CMP vs. CFUE)")
sum(res$padj<0.05)
dev.off()
```

```{r}
sig_regions <- res[which(res$padj < 0.05), ]
sig_regions <- merge(as.data.frame(sig_regions), 
                     as.data.frame(atac[, 1:3]), 
                     by.x = "row.names", 
                     by.y = "row.names", 
                     all.x = TRUE)

colnames(sig_regions)[1] <- "Region"
rownames(sig_regions) <- paste0(sig_regions$chrom, '_', sig_regions$start, '_', sig_regions$end)
head(sig_regions)
length(sig_regions[,1])

write.csv(sig_regions, "significant_atac_2vs3.csv", row.names = FALSE)
```

```{r}

upregulated <- sig_regions[sig_regions$log2FoldChange > 1, ]
head(upregulated)
length(upregulated[,1])
downregulated <- sig_regions[sig_regions$log2FoldChange < -1, ]
head(downregulated)
length(downregulated[,1])
```

```{r}

gene_symbols <- keys(org.Mm.eg.db, keytype = "SYMBOL")
gene_names <- AnnotationDbi::select(org.Mm.eg.db, keys = gene_symbols, columns = c("ENSEMBL", "GENENAME"), keytype = "SYMBOL")
dfPA <- gene_names[, c("ENSEMBL", "SYMBOL", "GENENAME")]
colnames(dfPA) <- c("geneId", "SYMBOL", "GENENAME")
head(dfPA)

```



```{r}
upregulated <- sig_regions[sig_regions$log2FoldChange > 1, ]
downregulated <- sig_regions[sig_regions$log2FoldChange < -1, ]

gr_up <- GRanges(seqnames = upregulated$chrom,
                 ranges = IRanges(start = upregulated$start - 3000, end = upregulated$end + 3000))
gr_down <- GRanges(seqnames = downregulated$chrom,
                   ranges = IRanges(start = downregulated$start - 3000, end = downregulated$end + 3000))

peak_anno_up <- annotatePeak(gr_up, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = "org.Mm.eg.db")
peak_anno_down <- annotatePeak(gr_down, TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, annoDb = "org.Mm.eg.db")

gene_symbols_up <- unique(as.data.frame(peak_anno_up)$SYMBOL)
gene_symbols_down <- unique(as.data.frame(peak_anno_down)$SYMBOL)

go_enrich_up <- enrichGO(
  gene = gene_symbols_up,
  OrgDb = org.Mm.eg.db,
  keyType = "SYMBOL",
  ont = "BP",
  pvalueCutoff = 0.05
)

go_enrich_down <- enrichGO(
  gene = gene_symbols_down,
  OrgDb = org.Mm.eg.db,
  keyType = "SYMBOL",
  ont = "BP",
  pvalueCutoff = 0.05
)

CairoPNG("GO_up_pair6.png", width = 700, height = 500)
dotplot(go_enrich_up,
        showCategory = 10,
        font.size = 12,
          title = "GO Enrichment (BP) - Upregulated Regions (CMP vs. CFUE)")
dev.off()

CairoPNG("GO_down_pair6.png", width = 700, height = 500)
dotplot(go_enrich_down,
        showCategory = 10,
        font.size = 12,
        title = "GO Enrichment (BP) - Downregulated Regions (CMP vs. CFUE)")
dev.off()

```

### Question 8

```{r}
# refer to: https://hds-sandbox.github.io/bulk_RNAseq_course/develop/05c_count_normalization.html

# Define design matrix
counts <- atac[, 4:11]
design <- data.frame(row.names = colnames(atac[, 4:11]),
                     celltype = factor(rep(c("HSC", "CMP", "CFUE", "Ery"), each = 2)))
# Create DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = design,
                              design = ~celltype)
dds <- DESeq(dds)
# Normalize data
cm <- data.frame(counts(dds, normalized=TRUE))
rownames(cm) <- paste0(atac$chrom, '_', atac$start, '_', atac$end)
```

```{r}
res_all = results(dds)
print(mcols(res_all, use.names=T))
```

```{r}

dist_matrix <- dist(t(assay(dds)))
hc <- hclust(dist_matrix)

dend <- as.dendrogram(hc)

new_labels <- c("HSC_1", "HSC_2", "CMP_1", "CMP_2", "CFUE_1", "CFUE_2", "ERY_1", "ERY_2")
labels(dend) <- new_labels

dend <- color_branches(dend, k = 4)  
dend <- set(dend, "labels_cex", 1.2)  
dend <- set(dend, "branches_lwd", 2)  

CairoPNG("tree_atac.png", width = 600, height = 500)
par(mar = c(5, 4, 4, 2))  
plot(dend, main = "ATAC-seq Sample Clustering", cex.main = 1.5)
dev.off()
```

```{r}


mat <- cm[which(res_all$padj < 0.01 & abs(res_all$log2FoldChange) > 1), ]
mat <- mat - rowMeans(mat)

rownames(mat) <- make.names(rownames(mat), unique = TRUE)
colnames(mat) <- c("HSC_1", "HSC_2", "CMP_1", "CMP_2", "CFUE_1", "CFUE_2", "ERY_1", "ERY_2")

CairoPNG("clustering_atac.png", width = 600, height = 500)
pheatmap(mat,
         scale = "row",
         main = "Hierarchical Tree (ATAC-seq)",
         annotation_col = NULL,
         show_colnames = TRUE,
         show_rownames = FALSE)
dev.off()
```