---
title: "RNA-ATAC-analysis"
author: "GA4"
date: "2025-04-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Library Prepare

```{r setup, include=FALSE}
set.seed(2025)
options(max.print = 1e7)

library(edgeR)
library(limma)
library(ggplot2)
library(pheatmap)
library(org.Mm.eg.db)
library(clusterProfiler)
library(tibble)
library(Cairo) 
library(dplyr)
library(GenomicRanges)
library(biomaRt)
library(AnnotationHub)
library(rtracklayer)
library(Hmisc)
library(ggrepel)
```

## 2. Load and Prepare the Data

```{r load-data}

# Load result files from rna and atac analysis
atac_data <- read.csv("~/atac_regions.csv")
rna_data <- read.csv("~/limma_rna_regions.csv")

#rna_data <- read.csv("~/deseq2_rna_regions.csv")
```

## 3. Question 6

```{r}
mouse <- useEnsembl(biomart = "genes", dataset = "mmusculus_gene_ensembl")
genes <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", 
                              "chromosome_name", "start_position", "end_position", "strand"),
               mart = mouse)

genes <- genes[genes$chromosome_name %in% gsub("chr", "", atac_data$chrom), ]
genes$chromosome_name <- paste0("chr", genes$chromosome_name)

```


```{r}

atac_gr <- GRanges(seqnames = atac_data$chrom,
                   ranges = IRanges(start = atac_data$start, end = atac_data$end))

gene_gr <- GRanges(seqnames = genes$chromosome_name,
                   ranges = IRanges(start = genes$start_position, end = genes$end_position),
                   gene_id = genes$ensembl_gene_id)

nearby <- distanceToNearest(atac_gr, gene_gr, ignore.strand = TRUE)

atac_data$gene_id <- NA
atac_data$gene_id[queryHits(nearby)] <- mcols(gene_gr[subjectHits(nearby)])$gene_id

```

```{r}

merged <- merge(atac_data, rna_data, by = "gene_id")

merged <- merged %>%
  mutate(
    quadrant = case_when(
      log2FoldChange > 0 & logFC > 0 ~ "Up-Up",
      log2FoldChange < 0 & logFC < 0 ~ "Down-Down",
      TRUE ~ "Other"
    )
  )

sig_points <- merged %>% filter(padj < 0.05 | adj.P.Val < 0.05)
non_sig_points <- merged %>% filter(!(padj < 0.05 | adj.P.Val < 0.05)) %>% sample_n(size = min(200, n()))
merged_sampled <- bind_rows(sig_points, non_sig_points)

quadrant_colors <- c("Up-Up" = "#E76F51", "Down-Down" = "#457B9D", "Other" = "gray80")

CairoPNG("ATAC_RNA_association.png", width = 1200, height = 1000, res = 150)
ggplot(merged_sampled, aes(x = log2FoldChange, y = logFC, color = quadrant)) +
  geom_point(alpha = 0.7, size = 1) +
  scale_color_manual(values = quadrant_colors) +
  theme_minimal(base_size = 14) +
  labs(
    x = "ATAC log2FoldChange",
    y = "RNA log2FoldChange",
    title = "ATAC vs RNA log2 Fold Change",
    color = "Quadrant"
  ) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "gray50") +
  geom_hline(yintercept = 0, linetype = "dashed", color = "gray50")
dev.off()
```

cor_result <- cor.test(merged$log2FoldChange, merged$logFC, method = "pearson")
cor_text <- paste0("r = ", round(cor_result$estimate, 2), 
                   "\np = ", formatC(cor_result$p.value, format = "e", digits = 2))


